[static runtime] Add NNC

bwasti opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44130 [static runtime] Add NNC**
* #44131 [static runtime] Split out graph preparation from runtime

Differential Revision: [D23604303](https://our.internmc.facebook.com/intern/diff/D23604303)

hlu1 commented on torch/csrc/jit/runtime/static/impl.cpp:
```diff
@@ -101,57 +104,105 @@ ProcessedNode::ProcessedNode(Node* node) : node_(node) {
+  for (size_t i = 0; i < size; i++) {
+    Value* v = node_->inputs()[i];
+    auto f = workspace.find(v);
+    TORCH_CHECK(
```
Use TORCH_DCHECK please.

hlu1 commented on torch/csrc/jit/runtime/static/impl.cpp:
```diff
@@ -51,6 +52,8 @@ StaticRuntime::StaticRuntime(const torch::jit::Module& m)
     }
   }
 
+  FuseTensorExprs(graph_);
```
This should be optional. Turning NNC on by default is too risky.

hlu1 commented on torch/csrc/jit/runtime/static/impl.cpp:
```diff
@@ -101,57 +104,105 @@ ProcessedNode::ProcessedNode(Node* node) : node_(node) {
       }
+    } else if (node_->kind() == Symbol::fromQualString("tensorexpr::Group")) {
+      uint64_t h = 0;
+      auto hash = [](uint64_t x) {
```
Is there any documentation on the hashing algorithm?

hlu1 reviewed (changes requested):
I think the NNC fuser should be default off and optional on, for now, until we can evaluate it for the use cases that we are targeting.

How does NNC handle dynamic shape?

bwasti committed:
Update on "[static runtime] Add NNC"


Differential Revision: [D23604303](https://our.internmc.facebook.com/intern/diff/D23604303)

[ghstack-poisoned]
